### PR TITLE
Reworked signal system, added support for Callable and Signal

### PR DESCRIPTION
**Pull request by reduz**
Both add a more abstract way to access callable and signal data, and both are added to the type system (and Variant).

Signal connections are now made by passing a Callable instead of object and method, this opens the door to a lot more custom and efficient ways to get the callbacks. Likewise, signals can be accessed from variants, and methods are now first class (only as a getter though).

Existing within-engine calls to connect and disconnect were renamed to connect_compat and disconnect_compat, they will eventually be modified to use method pointers for performance (and because they dont require a binding for the method), but this is left for step 2.

### CODE REVIEW

**Comment by Zireael07** [comment]
I hope we get a documentation example as soon as this lands?

**Comment by reduz** [comment]
@Zireael07 I dont think so, gdscript is being reworked a lot the next months, so there wont be doc until the syntax changes are done

File: core/register_core_types.cpp [hunk 1]
```diff
@@ -99,6 +99,9 @@ extern void unregister_variant_methods();
 
 void register_core_types() {
 
+	//consistency check
+	ERR_FAIL_COND(sizeof(Callable) > 16);
+
```
Comment (neikeq) [inline core/register_core_types.cpp]: An `static_assert` would be better to make it a compiler error instead.

**Review by neikeq — APPROVED** [review]
Looks good to me.

**Comment by akien-mga** [comment]
The Mono build will need some custom bindings, `mono_glue=yes` builds fail:
```
(CoreCompile target) -> 
  Generated/NativeCalls.cs(3220,94): error CS0246: The type or namespace name 'Callable' could not be found (are you missing a using directive or an assembly reference?) [/home/akien/Projects/godot/godot.git/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
  Generated/NativeCalls.cs(3223,95): error CS0246: The type or namespace name 'Callable' could not be found (are you missing a using directive or an assembly reference?) [/home/akien/Projects/godot/godot.git/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
  Generated/NativeCalls.cs(3226,95): error CS0246: The type or namespace name 'Callable' could not be found (are you missing a using directive or an assembly reference?) [/home/akien/Projects/godot/godot.git/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
  Generated/GodotObjects/Object.cs(471,45): error CS0246: The type or namespace name 'Callable' could not be found (are you missing a using directive or an assembly reference?) [/home/akien/Projects/godot/godot.git/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
  Generated/GodotObjects/Object.cs(481,47): error CS0246: The type or namespace name 'Callable' could not be found (are you missing a using directive or an assembly reference?) [/home/akien/Projects/godot/godot.git/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
  Generated/GodotObjects/Object.cs(490,48): error CS0246: The type or namespace name 'Callable' could not be found (are you missing a using directive or an assembly reference?) [/home/akien/Projects/godot/godot.git/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
```

Otherwise the rest should build fine now after I've rebased on master and fixed a few missing occurrences of old style `connect()`.

File: core/register_core_types.cpp (hunk 1 above)
Comment (reduz) [reply core/register_core_types.cpp]: Ah nice, I thought static assert was C++17 but there is a C++11 version. Can probably do a separate PR later changing a lot of those we have

**Comment by akien-mga** [comment]
Thanks!

File: core/callable.cpp [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+	if (p_method == StringName()) {
+		object = 0;
+		ERR_FAIL_MSG("Method argument to Callable constructor must be a non-empty string");
+	}
+
+	object = p_object;
+	method = p_method;
+}
+Callable::Callable(CallableCustom *p_custom) {
+	if (p_custom->referenced) {
```
Comment (vnen) [inline core/callable.cpp]: I don't understand this limitation. Why you can't create two Callable with the same custom? The equality function even try to compare them by pointer.

File: core/register_core_types.cpp (hunk 1 above)
Comment (kuruk-mm) [reply core/register_core_types.cpp]: You can use it #36385 or discard if you already did, I spend 30 minutes searching, and I didn't found more.

**Pull request merged**